### PR TITLE
Add trait methods to `alloy_network_primitives::TransactionResponse`

### PR DESCRIPTION
**Pull request by emhane**
## Motivation

Removing the optimism feature from `reth-rpc-types-compat` in a simple way, requires implementing the conversion from a signed transaction into an rpc transaction object, into the crates that define the associated types of `alloy_network::Network`. Otherwise, we are not able to access the additional OP fields for example.

Moving into `reth-optimsim-rpc` is not an option, since neither the trait for conversion nor the `<op_alloy_network::Optimism as Network>::TransactionResponse` type are defined in the `reth-optimsim-rpc` crate

## Solution

Adds trait methods to `alloy_network_primitives::TransactionResponse` for filling a `alloy_network_primitives::Transaction` from `alloy_consensus::Signed<impl alloy_consensus::Transaction>`.

## PR Checklist

- [ ] Added Tests
- [x] Added Documentation
- [x] Breaking changes

### CODE REVIEW

File: crates/network-primitives/src/traits.rs [hunk 1]
```diff
@@ -44,6 +45,22 @@ pub trait TransactionResponse {
...
+
+    /// Assemble from signed transaction.
+    fn fill(
+        signed_tx: Signed<impl alloy_consensus::Transaction>,
+        signer: Address,
+        block_hash: Option<B256>,
+        block_number: Option<u64>,
+        base_fee: Option<u64>,
+        transaction_index: Option<usize>,
+    ) -> Self;
```
Comment (mattsse) [inline crates/network-primitives/src/traits.rs:63]: unsure we can do this easily.

so I'd like to hold off on this for now

File: crates/network-primitives/src/traits.rs [hunk 2]
```diff
@@ -44,6 +45,22 @@ pub trait TransactionResponse {
     /// Input data
     #[doc(alias = "calldata")]
     fn input(&self) -> &Bytes;
+
+    /// Returns the gas price formatted for the RPC response.
+    fn gas_price(tx: &impl alloy_consensus::Transaction, base_fee: Option<u64>) -> u128;
+
+    /// Returns the max fee per gas.
+    fn max_fee_per_gas(tx: impl alloy_consensus::Transaction) -> Option<u128>;
```
Comment (mattsse) [inline crates/network-primitives/src/traits.rs:53]: these are a bit weird, unsure why they need to be part of this trait

**Review by mattsse — CHANGES_REQUESTED** [review]
let's do the additional `trait Transaction` first and think about the TransactionResponse fn(&impl Transaction) functions separately

**Comment by emhane** [comment]
> let's do the additional `trait Transaction` first and think about the TransactionResponse fn(&impl Transaction) functions separately

that's exactly how it is done. see base of prs. first pr: https://github.com/alloy-rs/alloy/pull/1172

**Comment by mattsse** [comment]
sorry I missed that pr

File: crates/network-primitives/src/traits.rs (hunk 2 above)
Comment (emhane) [reply crates/network-primitives/src/traits.rs:53]: so computation for l1 tx types can be reused by l2, much better than l2s doing copy paste

File: crates/network-primitives/src/traits.rs (hunk 1 above)
Comment (emhane) [reply crates/network-primitives/src/traits.rs:63]: wdym? this is needed to make progress in reth

**Comment by emhane** [comment]
re-request review @mattsse

File: crates/consensus/src/transaction/mod.rs
Comment (mattsse) [inline crates/consensus/src/transaction/mod.rs:1]: can we extract this to a separate pr?

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by emhane** [comment]
extracted yet another bit https://github.com/alloy-rs/alloy/pull/1198

File: crates/network-primitives/src/traits.rs [hunk 3]
```diff
@@ -107,6 +124,25 @@ impl<T: TransactionResponse> TransactionResponse for WithOtherFields<T> {
...
         self.inner.input()
     }
+
+    fn gas_price(tx: &impl alloy_consensus::Transaction, base_fee: Option<u64>) -> u128 {
+        T::gas_price(tx, base_fee)
+    }
+
+    fn max_fee_per_gas(tx: impl alloy_consensus::Transaction) -> Option<u128> {
+        T::max_fee_per_gas(tx)
+    }
```
Comment (mattsse) [inline crates/network-primitives/src/traits.rs:134]: these don't make much sense because the caller can just call them on tx directly

File: crates/network-primitives/src/traits.rs [hunk 4]
```diff
@@ -107,6 +124,25 @@ impl<T: TransactionResponse> TransactionResponse for WithOtherFields<T> {
...
+
+    fn fill(
+        signed_tx: Signed<impl alloy_consensus::Transaction>,
+        signer: Address,
+        block_hash: Option<B256>,
+        block_number: Option<u64>,
+        base_fee: Option<u64>,
+        transaction_index: Option<usize>,
+    ) -> Self {
```
Comment (mattsse) [inline crates/network-primitives/src/traits.rs:143]: this function is problematic because this now effectively freezes what additional fields can be included in a tx response.

**Review by mattsse — COMMENTED** [review]
we can't really solve this on the type trait, this is a server side problem because only there this type is created

File: crates/rpc-types-eth/src/transaction/mod.rs [hunk 5]
```diff
@@ -322,7 +322,66 @@ impl TransactionResponse for Transaction {
...
+            }
+        }
+    }
+
+    fn max_fee_per_gas(tx: impl alloy_consensus::Transaction) -> Option<u128> {
+        match TxType::try_from(tx.ty()).expect("should decode") {
+            TxType::Legacy | TxType::Eip2930 => None,
+            TxType::Eip1559 | TxType::Eip4844 | TxType::Eip7702 => Some(tx.max_fee_per_gas()),
+        }
+    }
```
Comment (emhane) [inline crates/rpc-types-eth/src/transaction/mod.rs:346]: this implementation varies from the impl on `alloy_consensus::Transaction`. this code is ported from reth, without modifying the logic @mattsse

File: crates/network-primitives/src/traits.rs (hunk 3 above)
Comment (emhane) [reply crates/network-primitives/src/traits.rs:134]: https://github.com/alloy-rs/alloy/pull/1173#discussion_r1732734183

File: crates/network-primitives/src/traits.rs (hunk 4 above)
Comment (emhane) [reply crates/network-primitives/src/traits.rs:143]: using `TransactionInfo` isn't possible due to circular dep, should I move `TransactionInfo` into this crate?

File: crates/rpc-types-eth/src/transaction/mod.rs (hunk 5 above)
Comment (emhane) [reply crates/rpc-types-eth/src/transaction/mod.rs:346]: https://github.com/paradigmxyz/reth/blob/e6dc947a3728637aa3b0f78c893e8f24dc468357/crates/rpc/rpc-types-compat/src/transaction/mod.rs#L41-L60

File: crates/network-primitives/src/traits.rs (hunk 4 above)
Comment (mattsse) [reply crates/network-primitives/src/traits.rs:143]: I think this would be reasonable actually

File: crates/rpc-types-eth/src/transaction/mod.rs [hunk 6]
```diff
@@ -319,7 +331,67 @@ impl TransactionResponse for Transaction {
...
+
+    fn fill(
+        signed_tx: Signed<impl alloy_consensus::Transaction>,
+        signer: Address,
+        tx_info: TransactionInfo,
+    ) -> Self {
```
Comment (mattsse) [inline crates/rpc-types-eth/src/transaction/mod.rs:361]: these functions won't work, because they constrain type creation, we can not lock this in in the trait function

**Review by mattsse — COMMENTED** [review]
converting types from consensus to rpc can't really be solved like this because this constrains it to `alloy_consensus::Transaction`

File: crates/rpc-types-eth/src/transaction/mod.rs (hunk 6 above)
Comment (emhane) [reply crates/rpc-types-eth/src/transaction/mod.rs:361]: how come?

**Comment by mattsse** [comment]
closing this, because we can't support the fill feature like this, because this conversion is type specific.

and the additional gas functions doesn't really make sense like this because this info should be encapsulated in the transaction itself, especially if we proceed with

https://github.com/alloy-rs/alloy/pull/1169
